Update OpenSSL build configuration based on Mbed (#3451)
openenclave/openenclave: docs/DesignDocs/CryptoLibConfiguration.md

# Configuration Parity Analysis between MbedTLS and OpenSSL Crypto Libraries

This document compares the build-time configuration of MbedTLS, which
has been extenstively reviewed when the library was ported to OE,
against that of OpenSSL, which OE recently starts to support. The goal of this

> analysis is to ensure the parity behween the two. For certain configurations

Reviewer: ```suggestion
analysis is to ensure the parity between the two. For certain configurations
```

that cannot achieve parity, the analysis also tries to reason out the gap.

*Note:* The following table only lists the options that are revelant to
the cryptographic algorithms. The table uses *CONFIG* or *!CONFIG* to indicate the
corresponding configuration is set or not.

> ## Parity analysis based on MbedlTLS

Reviewer 2: Great table, really helpful for doing the parity analysis! I would suggest a couple of changes to make it easier to make sense of:
- Break apart the option name from the configuration of that option OE uses to make it easier to parse because a `!` in mbedTLS column reflects a configuration state, and the `!` in the OpenSSL column reflects the parity setting which makes following what the comments refer to as "supported" unclear.
- Clarify which settings are OE specific-choices vs. mbedTLS defaults. We generally don't care so much about the latter cases being different.  
- Remove the OE comments about why some options are configured from the Comment column, and use that column just to talk about how we achieve parity behavior in OpenSSL, or what the resulting gap is if not. 
   - Alternatively, move the comments about the OE decisions to the OE configuration column so that it's clear what the comment specifically applies to.
- Highlight the rows where we have explicit mismatches in behavior for configurations OE has chosen, or are of concern. e.g. add a bold phrase that is easy to search like **No parity**. 

For example, consider something like:

MbedTLS option | OE configuration | Equivalent OpenSSL option | Comment
:---|:---|:---|:---|
MBEDTLS_CIPHER_MODE_CFB | Disabled (by OE as unsafe block cipher mode) | N/A | **No parity.** OpenSSL always supports CFB mode and cannot be disabled by OE. |
MBEDTLS_RSA_NO_CRT | Disabled (MbedTLS default) | N/A | Default OpenSSL behavior is to use Chinese Remainder Theorem in RSA, so while OpenSSL does not support disabling its use, it has parity behavior in OE with mbedTLS. |
MBEDTLS_KEY_EXCHANGE_PSK_ENABLED | Disabled (by OE as not providing perfect forward secrecy and not recommended for future use.) | OPENSSL_NO_PSK | OE will use the `no-psk` configuration to disable PSK in OpenSSL for parity. |
MBEDTLS_MD5_C | Enabled (MbedTLS default, OE doesn't override this because it is still commonly used for backwards compatibility) | !OPENSSL_NO_MD5 | OpenSSL supports MD5 by default, so we keep that behavior for parity. |

!MBEDTLS_ECP_DP_SECP224R1_ENABLED | N/A | Cannot be disabled on OpenSSL. |
MBEDTLS_ECP_DP_SECP256R1_ENABLED | N/A | p256 matches NSA's suite B. OpenSSL supports SECP 256r1 (i.e., X9.62 prime256v1) by default.  |
MBEDTLS_ECP_DP_SECP384R1_ENABLED | N/A | p384 matches NSA's suite B. OpenSSL supports SECP 384r1 by default. |
MBEDTLS_ECP_DP_SECP521R1_ENABLED | N/A | p521 matches NSA's suite B. OpenSSL supports SPEC 521r1 by default. |
!MBEDTLS_ECP_DP_SECP192K1_ENABLED | N/A | Cannot be disabled on OpenSSL. |
!MBEDTLS_ECP_DP_SECP224K1_ENABLED | N/A | Cannot be disabled on OpenSSL. |
MBEDTLS_ECP_DP_SECP256K1_ENABLED | N/A | p256k1 is used by bitcoin. OpenSSL supports SECP256k1 by default. |
!MBEDTLS_ECP_DP_BP256R1_ENABLED | N/A | Cannot be disabled on OpenSSL. |
!MBEDTLS_ECP_DP_BP384R1_ENABLED | N/A | Cannot be disabled on OpenSSL. |
!MBEDTLS_ECP_DP_BP512R1_ENABLED | N/A | Cannot be disabled on OpenSSL. |
MBEDTLS_ECP_DP_CURVE25519_ENABLED | N/A | OpenSSL supports curve 25519 by default. |
MBEDTLS_ECP_DP_CURVE448_ENABLED | N/A | OpenSSL supports curve 448 by default. |
MBEDTLS_ECP_NIST_OPTIM | N/A | No equivalent option in OpenSSL. |

> !MBEDTLS_ECP_RESTARTABLE | N/A | No equivalent option in OpenSSL. |

Reviewer 2: This is an mbedTLS specific feature, see https://tls.mbed.org/kb/development/restartable-ecc. I would just comment that mbedTLS defaults this to disabled and so OpenSSL has parity behavior anyway by not supporting it.

Author: I removed this configuration as I think it's implementation specific.

!MBEDTLS_KEY_EXCHANGE_DHE_RSA_ENABLED | N/A | Not supported in favor of ECDHE for performance. No equivalent option in OpenSSL. The following ciphersuites are supported by OpenSSL: `DHE_RSA_WITH_AES_256_GCM_SHA384`, `DHE_RSA_WITH_CAMELLIA_256_CBC_SHA256`, `DHE_RSA_WITH_CAMELLIA_256_CBC_SHA`, `DHE_RSA_WITH_AES_128_GCM_SHA256`, `DHE_RSA_WITH_CAMELLIA_128_CBC_SHA256`, and `DHE_RSA_WITH_CAMELLIA_128_CBC_SHA`. |
MBEDTLS_KEY_EXCHANGE_ECDHE_RSA_ENABLED | N/A | OpenSSL supports ECDHE-RSA based ciphersuites by default. No equivalent option in OpenSSL. |
MBEDTLS_KEY_EXCHANGE_ECDHE_ECDSA_ENABLED | N/A | OpenSSL supports ECDHE-ECDSA based ciphersuites by default. No equivalent option in OpenSSL. |
!MBEDTLS_KEY_EXCHANGE_ECDH_ECDSA_ENABLED | N/A | Drop uncommon protocol to minimize TCB. No equivalent option in OpenSSL. |
!MBEDTLS_KEY_EXCHANGE_ECDH_RSA_ENABLED | N/A | Drop uncommon protocol to minimize TCB. No equivalent option in OpenSSL. |

> !MBEDTLS_KEY_EXCHANGE_ECJPAKE_ENABLED | N/A | No equivalent option in OpenSSL. |

Reviewer 2: This should also comment that "OpenSSL does not support JPAKE (removed in v1.1.0)" which is a clearer explanation used for MBEDTLS_ECJPAKE_C

Author: Updated. Thanks for the information!

!MBEDTLS_PK_PARSE_EC_EXTENDED | N/A | Drop support for non-standard EC Curve. No equivalent option in OpenSSL. |
MBEDTLS_PKCS1_V15 | N/A | - |
MBEDTLS_PKCS1_V21 | N/A | - |
!MBEDTLS_RSA_NO_CRT | N/A | Cannot be disabled on OpenSSL. |

> MBEDTLS_AESNI_C | AES_ASM | - |

Reviewer 2: Shouldn't this be AESNI_ASM instead?

Author: OpenSSL's assembly implementation of AES uses AESNI instructions by default.

MBEDTLS_AES_C | N/A | - |
!MBEDTLS_ARC4_C | OPENSSL_NO_RC4 | Disable on OpenSSL with the no-rc4 configuration. |
MBEDTLS_BIGNUM_C | N/A | - |
!MBEDTLS_BLOWFISH_C | OPENSSL_NO_BF | Drop support for uncommon cipher. Disable on OpenSSL with the no-bf configuration. |
!MBEDTLS_CAMELLIA_C | OPENSSL_NO_CAMELLIA | Drop support for uncommon cipher. Disable on OpenSSL with the no-camellia configuration. |
!MBEDTLS_ARIA_C | OPENSSL_NO_ARIA | Drop support for uncommon cipher. Disable on OpenSSL with the no-aria configuration. |
MBEDTLS_CCM_C | N/A | - |
!MBEDTLS_CHACHA20_C | OPENSSL_NO_CHACHA | Disable less common cipher until there's a demand for it. Disable on OpenSSL with the no-chacha configuration. |
!MBEDTLS_CHACHAPOLY_C | OPENSSL_NO_CHACHA | Disable less common cipher until there's a demand for it. Disable on OpenSSL with the no-chacha configuration. |
MBEDTLS_CMAC_C | !OPENSSL_NO_CMAC | Enable as it's broadly used, allowed by NIST SP standards. |

> MBEDTLS_DES_C | !OPENSSL_NO_DES | Enable for backward compatibility as some protocols use it (e.g. payment industry protocols). |

Reviewer: ```suggestion
MBEDTLS_DES_C | !OPENSSL_NO_DES | Enable for backward compatibility as some protocols use it (e.g., payment industry protocols). |
```

MBEDTLS_ECDH_C | !OPENSSL_NO_EC | - |
MBEDTLS_ECDSA_C | !OPENSSL_NO_EC | - |
MBEDTLS_ECJPAKE_C | N/A | OpenSSL does not support JPAKE (removed in v1.1.0). |
MBEDTLS_ECP_C | N/A | - |
MBEDTLS_GCM_C | N/A | - |
!MBEDTLS_HAVEGE_C | N/A | Mbed-specific option. |
MBEDTLS_HKDF_C | N/A | OpenSSL supports HKDF by default (started from v1.1.0). |
MBEDTLS_HMAC_DRBG_C | N/A | Mbed-specific option. |
!MBEDTLS_NIST_KW_C | N/A | No equivalent option in OpenSSL. |
MBEDTLS_MD_C | N/A | No equivalent option in OpenSSL. |
!MBEDTLS_MD2_C | OPENSSL_NO_MD2 | Disable on OpenSSL with the no-md2 configuration. |
!MBEDTLS_MD4_C | OPENSSL_NO_MD4 | Disable on OpenSSL with the no-md4 configuration. |
MBEDTLS_MD5_C | !OPENSSL_NO_MD5 | Enable for backward compatibility. The algorithm is still commonly used. |
MBEDTLS_PKCS5_C | N/A | - |
!MBEDTLS_PKCS11_C | N/A | - |
MBEDTLS_PKCS12_C | N/A | - |

> !MBEDTLS_POLY1305_C | OPENSSL_NO_POLY1305 | Drop the uncommon hash algorithm to minimize TCB. Disable on OpenSSL with the no-poly1305 configuration.

Reviewer: ```suggestion
!MBEDTLS_POLY1305_C | OPENSSL_NO_POLY1305 | Drop the uncommon hash algorithm to minimize TCB. Disable on OpenSSL with the no-poly1305 configuration. |
```

!MBEDTLS_RIPEMD160_C | OPENSSL_NO_RMD160 | Drop the uncommon hash algorithm to minimize TCB. Disable on OpenSSL with the no-rmd160 configuration. |
MBEDTLS_RSA_C | !OPENSSL_NO_RSA | - |
MBEDTLS_SHA1_C | N/A | OpenSSL supports SHA1 by default. |
MBEDTLS_SHA256_C | N/A | OpenSSL supports SHA256 by default. |
MBEDTLS_SHA512_C | N/A | OpenSSL supports SHA512 by default. |

> !MBEDTLS_XTEA_C | N/A | No equivalent option in OpenSSL. |

Reviewer 2: Probably clearer to state that OpenSSL does not support the XTEA cipher, and so has parity with mbedTLS since OE disables this uncommon cipher.

Author: Updated.

## OpenSSL-specific Configuration

OpenSSL | Comment
|:---|:---|

> OPENSSL_NO_BLAKE2 | Blake2 hash is not supported by MbedTLS. Disable on OpenSSL with the no-blake2 configuration to minimize TCB. |

Reviewer: ```suggestion
OPENSSL_NO_BLAKE2 | Blake2 hash is not supported by Mbed TLS. Disable on OpenSSL with the no-blake2 configuration to minimize TCB. |
```
nit: In prose, there should be a space in "Mbed TLS" (see https://en.wikipedia.org/wiki/Mbed_TLS and https://github.com/ARMmbed/mbedtls).

OPENSSL_NO_CAST | CAST5 block cipher is not supported by MbedTLS. Disable on OpenSSL with the no-cast configuration to minimize TCB. |

> OPENSSL_NO_GOST | Russian GOST crypto engine is not supported by mbedTLS. Require dynamic loading and therefore not supported by the OE. Disable on OpenSSL with the no-gost configuration. |

Reviewer: ```suggestion
OPENSSL_NO_GOST | Russian GOST crypto engine is not supported by Mbed TLS. Require dynamic loading and therefore not supported by the OE. Disable on OpenSSL with the no-gost configuration. |
```
nit: either all lower case "mbedtls" like when you're referring to the github repo, or capitalized "Mbed TLS" in prose.
